File: Src2/Icm.Core/Functions/Auxiliary/FunctionPointPair.vb
Namespace Icm.Functions

    Public Class FunctionPointPair(Of TX As {Structure, IComparable(Of TX)}, TY As {Structure, IComparable(Of TY)})
        Inherits Tuple(Of FunctionPoint(Of TX, TY), FunctionPoint(Of TX, TY))

        ReadOnly Property X1() As TX
            Get
                Return Item1.X
            End Get
        End Property

        ReadOnly Property Y1() As TY
            Get
                Return Item1.Y
            End Get
        End Property

        ReadOnly Property X2() As TX
            Get
                Return Item2.X
            End Get
        End Property

        ReadOnly Property Y2() As TY
            Get
                Return Item2.Y
            End Get
        End Property

        ReadOnly Property P1() As FunctionPoint(Of TX, TY)
            Get
                Return Item1
            End Get
        End Property

        ReadOnly Property P2() As FunctionPoint(Of TX, TY)
            Get
                Return Item2
            End Get
        End Property

        Public Sub New(ByVal fun As IMathFunction(Of TX, TY))
            MyBase.New( _
            New FunctionPoint(Of TX, TY)(Nothing, fun), _
            New FunctionPoint(Of TX, TY)(Nothing, fun))
        End Sub

        Public Sub New(ByVal fun As IMathFunction(Of TX, TY), ByVal px1 As TX, ByVal px2 As TX)
            MyBase.New( _
            New FunctionPoint(Of TX, TY)(px1, fun), _
            New FunctionPoint(Of TX, TY)(px2, fun))
        End Sub

    End Class

End Namespace

File: Src/Icm.Core/Functions/Auxiliary/FunctionPointPair.cs
using System;

namespace Icm.Functions
{

	public class FunctionPointPair<TX, TY> : Tuple<FunctionPoint<TX, TY>, FunctionPoint<TX, TY>> where TX : struct, IComparable<TX> where TY : struct, IComparable<TY>
	{

		public TX X1 => Item1.X;
	    public TY Y1 => Item1.Y;
	    public TX X2 => Item2.X;
	    public TY Y2 => Item2.Y;
	    public FunctionPoint<TX, TY> P1 => Item1;
	    public FunctionPoint<TX, TY> P2 => Item2;

	    public FunctionPointPair(IMathFunction<TX, TY> fun) : base(
            new FunctionPoint<TX, TY>(default(TX), fun), 
            new FunctionPoint<TX, TY>(default(TX), fun))
		{
		}

		public FunctionPointPair(IMathFunction<TX, TY> fun, TX px1, TX px2) : base(
            new FunctionPoint<TX, TY>(px1, fun), 
            new FunctionPoint<TX, TY>(px2, fun))
		{
		}
	}
}